### DISCUSSION TITLE
Funky switch in Button Box

### DISCUSSION THREAD

**Discussion by alexdelli** [afpineda/OpenSourceSimWheelESP32#5] [Q&A] [answered]
Hi could you please help me for a layout to add a funky switch to the Button box setup 2?

thank you

**Comment by afpineda** [comment]
Will look into it. Give me some time, please.

**Comment by afpineda** [comment, marked as answer]
Thanks for your patience and support.
Modify the circuit in this way:

![Setup2_funky](https://github.com/afpineda/OpenSourceSimWheelESP32/assets/74291754/9c217ea6-43bd-4087-9de3-ee34eeb6682f)

- The funky switch replaces the fourth rotary encoder.
- The actual wiring of your funky switch **may not match** the one shown here. Look for a data sheet.
- **Misleading tags** have been reported at some ALPS funky switches. Be warned. Those show erroneously swapped `COM` and `PUSH` tags, at least, in their data sheet. Check first. If you come into one of them, swap those terminals.

The sketch needs some changes, too:

```c++
void simWheelSetup()
{
    inputs::addButtonMatrix(
        mtxSelectors,
        sizeof(mtxSelectors) / sizeof(mtxSelectors[0]),
        mtxInputs,
        sizeof(mtxInputs) / sizeof(mtxInputs[0]),
        mtxNumbers);
    inputs::addRotaryEncoder(GPIO_NUM_26, GPIO_NUM_27, 56, 57);
    inputs::addRotaryEncoder(GPIO_NUM_12, GPIO_NUM_13, 58, 59);
    inputs::addRotaryEncoder(GPIO_NUM_18, GPIO_NUM_19, 60, 61);
    inputs::addRotaryEncoder(GPIO_NUM_22, GPIO_NUM_23, 62, 63, true); // Funky switch
    inputHub::setDPADControls(10, 11, 12, 13);
}

void setup()
{
    userSettings::begin();
    simWheelSetup();
    hidImplementation::begin(
        DEVICE_NAME,
        DEVICE_MANUFACTURER,
        false);
    inputs::start();
}
```

I did some work for you:

[Setup2_funky.zip](https://github.com/afpineda/OpenSourceSimWheelESP32/files/14523995/Setup2_funky.zip)

1. Create a new folder named "Setup2_funky" (**case-sensitive**) at "your project root/src/Firmware".
2. Copy "Setup2_funky.ino" and "includes.txt" from the ZIP file into that folder:
3. Run the [source code setup procedure](https://github.com/afpineda/OpenSourceSimWheelESP32/blob/main/doc/firmware/sourcesSetup_en.md)
4. Open with Arduino IDE. 
5. Configure the board manager for "ESP32 Dev Module". Set "Core Debug Level" to "Error" in board configuration.
6. Upload.
7. Open the "serial monitor".
8. Reset.
9. Check that there are no error messages.

You have to do the testing, since I can not build this design by myself right now.
Please, send feedback and click "mark as answer" if it works.

Best regards.

**Reply by afpineda** [reply]
Sorry, there is a mistake. `COM` is not placed there in the funky switch. It is placed just below `ENCODER_COM` and `C (Down)`.

**Reply by alexdelli** [reply]
thank you mate, I'll test when I'll receive the funky switch from China and I'll give you the feedback
thank you for your work

**Comment by afpineda** [comment]
> thank you mate, I'll test when I'll receive the funky switch from China and I'll give you the feedback thank you for your work

Note: in the meanwhile, the firmware has gone to version 4.0.0. You need more code changes, now. Take a look to Setup2 again.
